predicators/ml_models.py

```diff
@@ -1303,9 +1306,87 @@ def _train_pytorch_model(model: nn.Module,
     return best_loss
 
 
+_D = TypeVar("_D")  # Datatype that will be stored in ReplayBuffer.
+
+
+class ReplayBuffer(Generic[_D]):
+    """A structure that holds data collected via online exploration.
+
+    This data will be used for model training.
+    """
+
+    def __init__(self, rng: np.random.Generator) -> None:
+        self._rng = rng
+
+    @abc.abstractmethod
+    def add_datum(self, datum: _D) -> None:
+        """Add a single datapoint to the replay buffer."""
+        raise NotImplementedError("Subclasses should override!")
+
+    @abc.abstractmethod
+    def sample_random_batch(self, batch_size: int) -> List[_D]:
+        """Return a batch of a specified size."""
+        raise NotImplementedError("Subclasses should override!")
+
+    @abc.abstractmethod
+    def dump_all_data(self) -> List[_D]:
+        """Return all data that is currently in the replay buffer."""
+        raise NotImplementedError("Subclasses should override!")
+
+
+class FixedSizeReplayBuffer(ReplayBuffer, Generic[_D]):
+    """A replay buffer implementation that has a fixed size.
+
+    If the replay buffer ever gets full, we will start replacing data
+    inside it.
+    """
+
+    def __init__(self, max_buffer_size: int, sample_with_replacement: bool,
+                 rng: np.random.Generator) -> None:
+        super().__init__(rng)
+        assert max_buffer_size > 0
+        self._max_buffer_size = max_buffer_size
+        self._buffer: List[_D] = []
+        self._sample_with_replacement = sample_with_replacement
+        self._curr_ptr_idx = 0
+
+    def add_datum(self, datum: _D) -> None:
+        assert len(self._buffer) <= self._max_buffer_size
+        # If the pointer is the same as the size of the buffer, then we
+        # know that we haven't yet filled the buffer, so we can
+        # simply append. Otherwise, we need to replace the current
+        # element.
+        if self._curr_ptr_idx == len(self._buffer):
+            self._buffer.append(datum)
+        else:
+            self._buffer[self._curr_ptr_idx] = datum
+        # Finally, we increment the pointer and handle wraparound
+        # correctly.
+        self._curr_ptr_idx += 1
+        self._curr_ptr_idx = self._curr_ptr_idx % self._max_buffer_size
+
+    def sample_random_batch(self, batch_size: int) -> List[_D]:
```

Comment:
wait hold on... is this function actually used?? I don't think that it is... and if we remove it, I'm wondering if we can literally remove the whole replay buffer class and just store a deque instead of the Q function itself? because at the moment, I think the only purpose of the replay buffer is to deque?

Reply:
you're right: the only purpose of the replay buffer is to deque. I realized this kinda late, but left the replay buffer implementation there in case we want to use it in the future for something else. But I guess it's better to just simplify things and then re-implement in the future if necessary. I can switch to using a queue.